File: platforms/ios/SparrowInvest/Models/PortfolioHistory.swift
//
//  PortfolioHistory.swift
//  SparrowInvest
//
//  Portfolio historical data for charts
//

import Foundation

// MARK: - Portfolio History Point

struct PortfolioHistoryPoint: Codable, Identifiable {
    let id: String
    let date: Date
    let value: Double
    let invested: Double

    var returns: Double {
        value - invested
    }

    var returnsPercentage: Double {
        guard invested > 0 else { return 0 }
        return ((value - invested) / invested) * 100
    }

    init(id: String = UUID().uuidString, date: Date, value: Double, invested: Double) {
        self.id = id
        self.date = date
        self.value = value
        self.invested = invested
    }
}

// MARK: - Portfolio History

struct PortfolioHistory: Codable {
    var dataPoints: [PortfolioHistoryPoint]
    var period: HistoryPeriod

    static var empty: PortfolioHistory {
        PortfolioHistory(dataPoints: [], period: .oneYear)
    }

    var minValue: Double {
        dataPoints.map { $0.value }.min() ?? 0
    }

    var maxValue: Double {
        dataPoints.map { $0.value }.max() ?? 0
    }

    var latestValue: Double {
        dataPoints.last?.value ?? 0
    }

    var earliestValue: Double {
        dataPoints.first?.value ?? 0
    }

    var periodReturn: Double {
        guard earliestValue > 0 else { return 0 }
        return ((latestValue - earliestValue) / earliestValue) * 100
    }
}

// MARK: - History Period

enum HistoryPeriod: String, Codable, CaseIterable {
    case oneMonth = "1M"
    case threeMonths = "3M"
    case sixMonths = "6M"
    case oneYear = "1Y"
    case threeYears = "3Y"
    case fiveYears = "5Y"
    case all = "All"

    var days: Int {
        switch self {
        case .oneMonth: return 30
        case .threeMonths: return 90
        case .sixMonths: return 180
        case .oneYear: return 365
        case .threeYears: return 365 * 3
        case .fiveYears: return 365 * 5
        case .all: return 365 * 10
        }
    }
}

File: platforms/android-consumer/app/src/main/java/com/sparrowinvest/app/data/model/PortfolioHistory.kt
package com.sparrowinvest.app.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class PortfolioHistoryPoint(
    val id: String,
    val date: String,
    val value: Double,
    val invested: Double
) {
    val returns: Double get() = value - invested
    val returnsPercentage: Double
        get() = if (invested > 0) ((value - invested) / invested) * 100 else 0.0
}

@Serializable
data class PortfolioHistory(
    @SerialName("data_points")
    val dataPoints: List<PortfolioHistoryPoint> = emptyList(),
    val period: HistoryPeriod = HistoryPeriod.ONE_YEAR
) {
    val minValue: Double get() = dataPoints.minOfOrNull { it.value } ?: 0.0
    val maxValue: Double get() = dataPoints.maxOfOrNull { it.value } ?: 0.0
    val latestValue: Double get() = dataPoints.lastOrNull()?.value ?: 0.0
    val earliestValue: Double get() = dataPoints.firstOrNull()?.value ?: 0.0

    val periodReturn: Double
        get() = if (earliestValue > 0) ((latestValue - earliestValue) / earliestValue) * 100 else 0.0

    companion object {
        val empty = PortfolioHistory()
    }
}

@Serializable
enum class HistoryPeriod(val label: String, val days: Int) {
    @SerialName("1M")
    ONE_MONTH("1M", 30),

    @SerialName("3M")
    THREE_MONTHS("3M", 90),

    @SerialName("6M")
    SIX_MONTHS("6M", 180),

    @SerialName("1Y")
    ONE_YEAR("1Y", 365),

    @SerialName("3Y")
    THREE_YEARS("3Y", 365 * 3),

    @SerialName("5Y")
    FIVE_YEARS("5Y", 365 * 5),

    @SerialName("All")
    ALL("All", 365 * 10);
}
